### PR TITLE
issue #29638 - Improvements to contacts list

### PR DESCRIPTION
**Pull request by ChristopherCotnoir**

### CODE REVIEW

**Comment by ChristopherCotnoir** [comment]
Requires xtuple/xtuple#3196.

File: guiclient/contact.cpp [hunk 1]
```diff
@@ -591,6 +591,44 @@ void contact::sSave()
...
+    _notes->setEnabled(false);
+    _comments->setReadOnly(true);
+    _documents->setReadOnly(true);
+    _charass->setReadOnly(true);
+  }
+
+  QWidgetList list = QApplication::allWidgets();
+  _close = false;
+
+  for (int i = 0; i < list.size(); i++)
```
Comment (gilmoskowitz) [inline guiclient/contact.cpp]: Nice idea here.
Consider simplifying with `foreach(QWidget *widget, QApplication::allWidgets()) {...}`.

File: guiclient/contact.h [hunk 2]
```diff
@@ -30,6 +31,8 @@ class contact : public XDialog, public Ui::contact
     contact(QWidget* parent = 0, const char* name = 0, bool modal = false, Qt::WindowFlags fl = 0);
     ~contact();
 
+    Q_INVOKABLE virtual int id() { return _cntctid; }
```
Comment (gilmoskowitz) [inline guiclient/contact.h]: Try to avoid function code in header files.

File: guiclient/contact.h [hunk 3]
```diff
@@ -80,12 +83,16 @@ protected slots:
...
+    virtual void setVisible(bool);
+
 protected:
     contactPrivate *_data;
     virtual void closeEvent(QCloseEvent*);
 
 private:
     int _cntctid;
+    bool _close;
+    AppLock _lock;
```
Comment (gilmoskowitz) [inline guiclient/contact.h]: Consider putting these in `contactPrivate` instead of directly in `contact`.

File: guiclient/contacts.cpp [hunk 4]
```diff
@@ -194,34 +185,54 @@ void contacts::sView()
...
+      bool cascade = false;
+      chk.bindValue(":cntct_id", selected->id());
+      chk.exec();
+      if (chk.first() && chk.value("inUse").toBool())
+      {
+        if (QMessageBox::warning(this, tr("Delete Parent Objects?"),
+                                 tr("There are parent objects that use this contact. Do you wish "
+                                    "to delete these objects as well?"),
+                                 QMessageBox::Yes,
+                                 QMessageBox::No | QMessageBox::Default) == QMessageBox::Yes)
```
Comment (gilmoskowitz) [inline guiclient/contacts.cpp]: This is a deprecated version of `QMessageBox::warning()` (same above).

**Pull request merged**